### PR TITLE
CSCQVAIN-324: ADD: logging functionality in case of error

### PR DESCRIPTION
**Pull request by ShreyasDeshpande**
Adds error_id for an json passed on the frontend. logs the information at console

### CODE REVIEW

File: cmd/qvain-backend/api_helpers.go [hunk 1]
```diff
@@ -146,50 +147,49 @@ func apiVersion(w http.ResponseWriter, r *http.Request) {
 
 // dbError handles database errors. It returns more specific API messages for predefined errors
 // that might be relevant for the user. Other errors return `database error` with a 500 status code.
-func dbError(w http.ResponseWriter, err error) bool {
+// Also loggs error message to backend terminal
```
Comment (jppiiroinen) [inline cmd/qvain-backend/api_helpers.go]: "loggs"

File: cmd/qvain-backend/dataset_api.go [hunk 2]
```diff
@@ -119,7 +117,7 @@ func (api *DatasetApi) ListDatasets(w http.ResponseWriter, r *http.Request, user
 
 // Dataset handles requests for a dataset by UUID. It dispatches to request method specific handlers.
 func (api *DatasetApi) Dataset(w http.ResponseWriter, r *http.Request, user *models.User, id uuid.UUID) {
-	//api.logger.Debug().Str("head", "").Str("path", r.URL.Path).Msg("dataset")
+	api.logger.Debug().Str("head", "").Str("path", r.URL.Path).Msg("dataset")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/dataset_api.go:120]: why commented in?

File: cmd/qvain-backend/dataset_api.go [hunk 3]
```diff
@@ -142,7 +140,7 @@ func (api *DatasetApi) Dataset(w http.ResponseWriter, r *http.Request, user *mod
 			return
 
 		default:
-			jsonError(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+			loggedJSONError(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed, &api.logger).Msg("Error in Dataset method dispach")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/dataset_api.go]: "Error in Dataset method dispach" This does not bring value on the logs, it is too large. It would make more sense to have some kind of NAME_001 style of line in the log itself, which would not consume space and would be easy to find.

**Review by jppiiroinen — CHANGES_REQUESTED** [review]

File: cmd/qvain-backend/dataset_api.go (hunk 3 above)
Comment (ShreyasDeshpande) [reply cmd/qvain-backend/dataset_api.go]: The Error message prints the line number of the error. so I think it should be easy to find. Main intention was to log the method information if possible in the error message.

File: cmd/qvain-backend/dataset_api.go (hunk 2 above)
Comment (ShreyasDeshpande) [reply cmd/qvain-backend/dataset_api.go:120]: It was commented for some reason, don't know why. but thought it would be helpful to log the URL path of the request coming in for dataset operation.so did uncomment

File: cmd/qvain-backend/api_helpers.go (hunk 1 above)
Comment (ShreyasDeshpande) [reply cmd/qvain-backend/api_helpers.go]: changed

File: cmd/qvain-backend/dataset_api.go (hunk 3 above)
Comment (jppiiroinen) [reply cmd/qvain-backend/dataset_api.go]: I would say that it is kind of an obvious that it is an error, so why to write "Error in" in there. Also would not for example "Unhandled request method (METHOD_HERE)" be more describing. The same applies to all default switch-cases.

File: cmd/qvain-backend/api_helpers.go [hunk 4]
```diff
@@ -146,50 +147,49 @@ func apiVersion(w http.ResponseWriter, r *http.Request) {
 
 // dbError handles database errors. It returns more specific API messages for predefined errors
 // that might be relevant for the user. Other errors return `database error` with a 500 status code.
-func dbError(w http.ResponseWriter, err error) bool {
+// Also logs error message to backend terminal
+func dbError(w http.ResponseWriter, err error, logger *zerolog.Logger) *zerolog.Event {
 	switch err {
 	case nil:
-		return false
+		return nil
```
Comment (jppiiroinen) [inline cmd/qvain-backend/api_helpers.go:154]: why this change?

Comment (ShreyasDeshpande) [reply cmd/qvain-backend/api_helpers.go:154]: we are returning object instead of bool (which it was earlier)

File: cmd/qvain-backend/session_api.go [hunk 5]
```diff
@@ -29,7 +29,7 @@ func (api *SessionApi) Current(w http.ResponseWriter, r *http.Request) {
 	session, err := api.sessions.SessionFromRequest(r)
 	if err != nil {
 		api.logger.Debug().Err(err).Msg("no current session")
-		sessionError(w, sessions.ErrSessionNotFound)
+		sessionError(w, sessions.ErrSessionNotFound, &api.logger)
```
Comment (jppiiroinen) [inline cmd/qvain-backend/session_api.go]: should the line above be connected to this one?

File: cmd/qvain-backend/api_proxy.go [hunk 6]
```diff
@@ -223,43 +222,43 @@ func (api *ApiProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 		if len(session.User.Projects) < 1 {
-			jsonError(w, "access denied: user has no projects", http.StatusForbidden)
+			loggedJSONError(w, "access denied: user has no projects", http.StatusForbidden, &api.logger).Msg("error in api_proxy")
 			return
 		}
 		project := projectQueries[0]
 		if !session.User.HasProject(project) {
 			api.logger.Debug().Strs("projects", session.User.Projects).Str("wanted", project).Msg("project check")
-			jsonError(w, "access denied: invalid project", http.StatusForbidden)
+			loggedJSONError(w, "access denied: invalid project", http.StatusForbidden, &api.logger).Msg("error in api_proxy")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/api_proxy.go]: invalid error message.. should be combined with the upper one.

Comment (jppiiroinen) [reply cmd/qvain-backend/api_proxy.go]: I wonder if we need to have this "error in api_proxy" at all, what if we would just enable the "caller" field in the logs?

https://github.com/rs/zerolog/commit/27e0a22cbc6f6a8f2a7ac2ed2c9b410b05f59e02

**Comment by jppiiroinen** [comment]
I wonder in general, that I think that we should write the actual errors into syslog/journald etc. So that we would have those even if the service will restart and also that those will be stored somewhere.

File: cmd/qvain-backend/dataset_api.go [hunk 7]
```diff
@@ -140,7 +140,7 @@ func (api *DatasetApi) Dataset(w http.ResponseWriter, r *http.Request, user *mod
 			return
 
 		default:
-			loggedJSONError(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed, &api.logger).Msg("Error in Dataset method dispach")
+			loggedJSONError(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed, &api.logger).Str("Resuest Method", r.Method).Msg("Unhandled request method")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/dataset_api.go]: "Resuest"

File: cmd/qvain-backend/dataset_api.go [hunk 8]
```diff
@@ -149,7 +149,7 @@ func (api *DatasetApi) Dataset(w http.ResponseWriter, r *http.Request, user *mod
 	switch op {
 	case "export":
 		// TODO: assess security implementations before enabling this
-		loggedJSONError(w, "export not implemented", http.StatusNotImplemented, &api.logger).Msg("Error in Dataset method dispatch")
+		loggedJSONError(w, "export not implemented", http.StatusNotImplemented, &api.logger).Msg("export failed")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/dataset_api.go]: "export failed" => "export not implemented"

File: cmd/qvain-backend/dataset_api.go [hunk 9]
```diff
@@ -191,20 +191,20 @@ func (api *DatasetApi) createDataset(w http.ResponseWriter, r *http.Request, cre
 	var err error
 
 	if !strings.HasPrefix(r.Header.Get("Content-Type"), "application/json") {
-		loggedJSONError(w, http.StatusText(http.StatusUnsupportedMediaType), http.StatusUnsupportedMediaType, &api.logger).Msg("Error in create Dataset")
+		loggedJSONError(w, http.StatusText(http.StatusUnsupportedMediaType), http.StatusUnsupportedMediaType, &api.logger).Msg("Dataset creation failed")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/dataset_api.go]: "Dataset creation failed" => "Unsupported content-type "
also it would be useful to get that value of r.Header.Get("Content-Type") to log

File: cmd/qvain-backend/dataset_api.go [hunk 10]
```diff
@@ -191,20 +191,20 @@ func (api *DatasetApi) createDataset(w http.ResponseWriter, r *http.Request, cre
...
 		return
 	}
 
 	if r.Body == nil || r.Body == http.NoBody {
-		loggedJSONError(w, "empty body", http.StatusBadRequest, &api.logger).Msg("Error in create Dataset")
+		loggedJSONError(w, "empty body", http.StatusBadRequest, &api.logger).Msg("Dataset creation failed")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/dataset_api.go]: "Dataset creation failed, because request body was empty."

File: cmd/qvain-backend/dataset_api.go (hunk 9 above)
Comment (jppiiroinen) [reply cmd/qvain-backend/dataset_api.go]: but that size needs to be limited.

File: cmd/qvain-backend/dataset_api.go (hunk 7 above)
Comment (jppiiroinen) [reply cmd/qvain-backend/dataset_api.go]: the length of r.Method should be ensured to be maximum of some length.

File: cmd/qvain-backend/dataset_api.go [hunk 11]
```diff
@@ -92,21 +92,21 @@ func (api *DatasetApi) ListDatasets(w http.ResponseWriter, r *http.Request, user
...
+			loggedJSONError(w, err.Error(), http.StatusBadRequest, &api.logger).Err(err).Msg("Listing dataset failed")
 			//dbError(w, err)
 			return
 		}
 	case "fetchall":
 		api.logger.Debug().Str("op", "fetchall").Msg("datasets")
 		shared.FetchAll(api.metax, api.db, api.logger, user.Uid, user.Identity)
 	default:
-		loggedJSONError(w, "invalid parameter", http.StatusBadRequest, &api.logger).Msg("Error while listing Datasets")
+		loggedJSONError(w, "invalid parameter", http.StatusBadRequest, &api.logger).Str("URL.RawQuery", r.URL.RawQuery).Msg("Unhandled parameter")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/dataset_api.go]: the length of r.URL.RawQuery should be ensured to be maximum of some length.

File: cmd/qvain-backend/dataset_api.go [hunk 12]
```diff
@@ -162,7 +162,7 @@ func (api *DatasetApi) Dataset(w http.ResponseWriter, r *http.Request, user *mod
 		}
 		return
 	default:
-		loggedJSONError(w, "invalid dataset operation", http.StatusNotFound, &api.logger).Msg("Error in Dataset method dispatch")
+		loggedJSONError(w, "invalid dataset operation", http.StatusNotFound, &api.logger).Str("op", op).Msg("Unhandled dataset operation")
```
Comment (jppiiroinen) [inline cmd/qvain-backend/dataset_api.go]: the length of `op` should be ensured to be maximum of some length.

**Review by jppiiroinen — CHANGES_REQUESTED** [review]

**Review by jppiiroinen — APPROVED** [review]
